File: lib/ruby2js/isr.rb
module Ruby2JS
  # ISR (Incremental Static Regeneration) adapter interface.
  #
  # ISR enables pages to be statically generated at runtime and cached,
  # with automatic background regeneration when stale. This provides the
  # performance of static sites with the freshness of dynamic content.
  #
  # Platform-specific adapters (Vercel, Cloudflare) implement this interface
  # to leverage each platform's caching infrastructure.
  #
  # == Usage in pages
  #
  # Use the revalidate pragma to enable ISR for a page:
  #
  #   # Pragma: revalidate 60
  #
  #   @posts = Post.all
  #   __END__
  #   <ul>
  #     <% @posts.each do |post| %>
  #       <li><%= post.title %></li>
  #     <% end %>
  #   </ul>
  #
  # The page will be cached for 60 seconds, then regenerated in the background
  # while serving stale content.
  #
  # == Implementing an adapter
  #
  #   class MyPlatformISR < Ruby2JS::ISR::Base
  #     def self.serve(context, cache_key, options = {}, &block)
  #       # Check cache, serve if fresh
  #       # If stale, serve stale and regenerate in background
  #       # If miss, generate, cache, and serve
  #     end
  #
  #     def self.revalidate(path)
  #       # Invalidate cache for the given path
  #     end
  #   end
  #
  module ISR
    # Default revalidation time in seconds (1 minute)
    DEFAULT_REVALIDATE = 60

    # Default stale-while-revalidate window (24 hours)
    DEFAULT_STALE_WHILE_REVALIDATE = 86400

    # Base class for ISR adapters.
    #
    # Platform-specific adapters should inherit from this class and implement
    # the +serve+ and +revalidate+ methods.
    #
    class Base
      # Serve content with ISR caching.
      #
      # @param context [Object] Request context (platform-specific)
      # @param cache_key [String] Cache key, typically the URL path
      # @param options [Hash] Caching options
      # @option options [Integer] :revalidate Seconds until content is stale (default: 60)
      # @option options [Integer] :stale_while_revalidate Seconds to serve stale content
      #   while regenerating (default: 86400)
      # @option options [Hash] :tags Cache tags for grouped invalidation
      # @yield Block that renders the content
      # @yieldparam context [Object] The request context
      # @yieldreturn [String] The rendered HTML content
      # @return [Object] Platform-specific response object
      #
      # @example
      #   ISRAdapter.serve(request, '/posts', revalidate: 60) do |ctx|
      #     render_posts_page(ctx)
      #   end
      #
      def self.serve(context, cache_key, options = {}, &block)
        raise NotImplementedError, "#{name}.serve must be implemented by subclass"
      end

      # Invalidate cached content for a path.
      #
      # Use this for on-demand revalidation when content changes, rather than
      # waiting for the revalidation window.
      #
      # @param path [String] The path to invalidate
      # @return [Boolean] true if invalidation was successful
      #
      # @example
      #   # After updating a post
      #   ISRAdapter.revalidate('/posts')
      #   ISRAdapter.revalidate("/posts/#{post.id}")
      #
      def self.revalidate(path)
        raise NotImplementedError, "#{name}.revalidate must be implemented by subclass"
      end

      # Invalidate cached content by tag.
      #
      # Some platforms support cache tags for grouped invalidation.
      # This is optional - adapters that don't support tags can skip this.
      #
      # @param tag [String] The cache tag to invalidate
      # @return [Boolean] true if invalidation was successful
      #
      # @example
      #   # Invalidate all posts-related caches
      #   ISRAdapter.revalidate_tag('posts')
      #
      def self.revalidate_tag(tag)
        raise NotImplementedError, "#{name}.revalidate_tag is not supported"
      end

      # Parse ISR options from pragma comments.
      #
      # @param comments [Array<String>] Comment lines from source
      # @return [Hash, nil] Parsed options or nil if no ISR pragma found
      #
      # @example
      #   parse_pragma(['# Pragma: revalidate 60'])
      #   # => { revalidate: 60 }
      #
      def self.parse_pragma(comments)
        comments.each do |comment|
          text = comment.is_a?(String) ? comment : comment.to_s
          if text =~ /Pragma:\s*revalidate\s+(\d+)/i
            return { revalidate: $1.to_i }
          end
        end
        nil
      end

      # Build Cache-Control header value for ISR.
      #
      # @param options [Hash] ISR options
      # @option options [Integer] :revalidate Seconds until stale
      # @option options [Integer] :stale_while_revalidate Stale serve window
      # @return [String] Cache-Control header value
      #
      def self.cache_control_header(options = {})
        revalidate = options[:revalidate] || DEFAULT_REVALIDATE
        stale_window = options[:stale_while_revalidate] || DEFAULT_STALE_WHILE_REVALIDATE

        "s-maxage=#{revalidate}, stale-while-revalidate=#{stale_window}"
      end
    end

    # In-memory ISR adapter for development and testing.
    #
    # This adapter provides a simple in-memory cache that mimics ISR behavior
    # without requiring a platform-specific caching layer.
    #
    # NOT suitable for production use - cache is not shared across processes
    # and has no size limits.
    #
    class Memory < Base
      @cache = {}
      @cache_times = {}

      class << self
        # Clear the in-memory cache (useful for testing)
        def clear_cache
          @cache = {}
          @cache_times = {}
        end

        def serve(context, cache_key, options = {}, &block)
          revalidate = options[:revalidate] || DEFAULT_REVALIDATE
          now = Time.now.to_i

          cached = @cache[cache_key]
          cached_time = @cache_times[cache_key]

          if cached && cached_time
            age = now - cached_time
            if age < revalidate
              # Fresh - serve from cache
              return cached
            else
              # Stale - serve stale and regenerate
              # In a real adapter, regeneration would happen in background
              # For memory adapter, we regenerate synchronously
              content = block.call(context)
              @cache[cache_key] = content
              @cache_times[cache_key] = now
              return cached  # Return stale content (simulating async regeneration)
            end
          end

          # Cache miss - generate and cache
          content = block.call(context)
          @cache[cache_key] = content
          @cache_times[cache_key] = now
          content
        end

        def revalidate(path)
          @cache.delete(path)
          @cache_times.delete(path)
          true
        end

        def revalidate_tag(tag)
          # Memory adapter doesn't support tags - clear everything
          clear_cache
          true
        end
      end
    end
  end
end

File: packages/ruby2js-rails/targets/cloudflare/isr.mjs
// Ruby2JS ISR (Incremental Static Regeneration) Adapter for Cloudflare Workers
//
// Uses Cloudflare's Cache API for edge caching with background regeneration
// via waitUntil(). Provides stale-while-revalidate behavior at the edge.
//
// Usage:
//   import { ISRCache } from 'ruby2js-rails/targets/cloudflare/isr.mjs';
//
//   export default {
//     async fetch(request, env, ctx) {
//       return ISRCache.serve({ request, env, ctx }, async (context) => {
//         const posts = await Post.all();
//         return renderPosts(posts);
//       }, { revalidate: 60 });
//     }
//   };

// Default revalidation time in seconds (1 minute)
const DEFAULT_REVALIDATE = 60;

// Default stale-while-revalidate window (24 hours)
const DEFAULT_STALE_WHILE_REVALIDATE = 86400;

/**
 * ISR Cache adapter for Cloudflare Workers.
 *
 * Uses Cloudflare's Cache API to store and retrieve responses at the edge.
 * Implements stale-while-revalidate pattern using waitUntil() for background
 * regeneration without blocking the response.
 *
 * @see https://developers.cloudflare.com/workers/runtime-apis/cache/
 */
export class ISRCache {
  /**
   * Serve content with ISR caching.
   *
   * @param {Object} context - Worker context
   * @param {Request} context.request - The incoming request
   * @param {Object} context.env - Worker environment bindings
   * @param {Object} context.ctx - Execution context with waitUntil
   * @param {Function} renderFn - Async function that renders content: (context) => Promise<string>
   * @param {Object} options - Caching options
   * @param {number} [options.revalidate=60] - Seconds until content is considered stale
   * @param {string} [options.cacheKey] - Custom cache key (defaults to request URL)
   * @param {Object} [options.headers={}] - Additional response headers
   * @returns {Promise<Response>} - Cached or freshly rendered response
   *
   * @example
   * // Basic usage
   * return ISRCache.serve({ request, env, ctx }, async () => {
   *   const posts = await Post.all();
   *   return `<ul>${posts.map(p => `<li>${p.title}</li>`).join('')}</ul>`;
   * }, { revalidate: 60 });
   *
   * @example
   * // With custom cache key
   * return ISRCache.serve({ request, env, ctx }, async (context) => {
   *   return renderPage(context);
   * }, { revalidate: 300, cacheKey: '/posts/all' });
   */
  static async serve(context, renderFn, options = {}) {
    const { request, ctx } = context;
    const revalidate = options.revalidate ?? DEFAULT_REVALIDATE;

    // Get the default cache
    const cache = caches.default;

    // Build cache key - use custom key or request URL
    const cacheKeyUrl = options.cacheKey
      ? new URL(options.cacheKey, request.url).href
      : request.url;
    const cacheKey = new Request(cacheKeyUrl, { method: 'GET' });

    // Check cache for existing response
    let response = await cache.match(cacheKey);

    if (response) {
      // Check if response is stale based on Age header
      const age = parseInt(response.headers.get('age') || '0', 10);
      const cacheTime = parseInt(response.headers.get('x-cache-time') || '0', 10);
      const now = Math.floor(Date.now() / 1000);
      const responseAge = cacheTime ? (now - cacheTime) : age;

      if (responseAge < revalidate) {
        // Fresh - serve from cache
        return this.addCacheHeaders(response.clone(), 'HIT', responseAge);
      }

      // Stale - serve stale content and regenerate in background
      if (ctx && ctx.waitUntil) {
        ctx.waitUntil(this.regenerate(context, cacheKey, renderFn, options));
      }
      return this.addCacheHeaders(response.clone(), 'STALE', responseAge);
    }

    // Cache miss - generate, cache, and serve
    return await this.regenerate(context, cacheKey, renderFn, options);
  }

  /**
   * Regenerate content and update cache.
   *
   * @param {Object} context - Worker context
   * @param {Request} cacheKey - Cache key request
   * @param {Function} renderFn - Render function
   * @param {Object} options - Caching options
   * @returns {Promise<Response>} - Fresh response
   */
  static async regenerate(context, cacheKey, renderFn, options = {}) {
    const { ctx } = context;
    const revalidate = options.revalidate ?? DEFAULT_REVALIDATE;
    const staleWhileRevalidate = options.staleWhileRevalidate ?? DEFAULT_STALE_WHILE_REVALIDATE;

    // Render fresh content
    const html = await renderFn(context);

    // Build response headers
    const headers = new Headers(options.headers || {});
    headers.set('Content-Type', 'text/html; charset=utf-8');
    headers.set('Cache-Control', `s-maxage=${revalidate}, stale-while-revalidate=${staleWhileRevalidate}`);
    headers.set('x-cache-time', String(Math.floor(Date.now() / 1000)));
    headers.set('x-cache-status', 'MISS');

    // Add cache tags if provided
    if (options.tags && Array.isArray(options.tags)) {
      headers.set('Cache-Tag', options.tags.join(','));
    }

    const response = new Response(html, {
      status: 200,
      headers,
    });

    // Store in cache (non-blocking)
    const cache = caches.default;
    if (ctx && ctx.waitUntil) {
      ctx.waitUntil(cache.put(cacheKey, response.clone()));
    } else {
      // Fallback: await the cache put
      await cache.put(cacheKey, response.clone());
    }

    return response;
  }

  /**
   * On-demand revalidation for a specific path.
   *
   * Deletes the cached response for the given path, forcing regeneration
   * on the next request.
   *
   * @param {string} path - The path to revalidate (full URL or path)
   * @param {Object} [options] - Options
   * @param {string} [options.baseUrl] - Base URL for relative paths
   * @returns {Promise<boolean>} - true if cache entry was deleted
   *
   * @example
   * // After updating a post
   * await ISRCache.revalidate('https://example.com/posts/123');
   *
   * @example
   * // With base URL
   * await ISRCache.revalidate('/posts/123', { baseUrl: 'https://example.com' });
   */
  static async revalidate(path, options = {}) {
    try {
      const cache = caches.default;

      // Build full URL if path is relative
      let url;
      if (path.startsWith('http://') || path.startsWith('https://')) {
        url = path;
      } else if (options.baseUrl) {
        url = new URL(path, options.baseUrl).href;
      } else {
        // Can't revalidate without knowing the full URL
        console.warn('ISRCache.revalidate: path must be a full URL or baseUrl must be provided');
        return false;
      }

      const cacheKey = new Request(url, { method: 'GET' });
      const deleted = await cache.delete(cacheKey);

      return deleted;
    } catch (error) {
      console.error('ISRCache.revalidate error:', error);
      return false;
    }
  }

  /**
   * Revalidate multiple paths at once.
   *
   * @param {string[]} paths - Array of paths to revalidate
   * @param {Object} [options] - Options (same as revalidate)
   * @returns {Promise<boolean[]>} - Array of results
   */
  static async revalidateMany(paths, options = {}) {
    return Promise.all(paths.map(path => this.revalidate(path, options)));
  }

  /**
   * Purge all cached content.
   *
   * Note: Cloudflare's Cache API doesn't support listing/purging all entries.
   * This method is provided for interface compatibility but has limited functionality.
   * Use Cloudflare's dashboard or API for full cache purging.
   *
   * @returns {Promise<boolean>} - Always returns false (not supported)
   */
  static async purgeAll() {
    console.warn('ISRCache.purgeAll: Not supported by Cloudflare Cache API. Use Cloudflare dashboard or API.');
    return false;
  }

  /**
   * Add cache status headers to response.
   *
   * @param {Response} response - Original response
   * @param {string} status - Cache status (HIT, MISS, STALE)
   * @param {number} [age] - Age in seconds
   * @returns {Response} - Response with cache headers
   */
  static addCacheHeaders(response, status, age) {
    const headers = new Headers(response.headers);
    headers.set('x-cache-status', status);
    if (age !== undefined) {
      headers.set('age', String(age));
    }

    return new Response(response.body, {
      status: response.status,
      statusText: response.statusText,
      headers,
    });
  }

  /**
   * Build Cache-Control header value.
   *
   * @param {Object} options - Caching options
   * @param {number} [options.revalidate] - s-maxage value
   * @param {number} [options.staleWhileRevalidate] - stale-while-revalidate value
   * @returns {string} - Cache-Control header value
   */
  static cacheControlHeader(options = {}) {
    const revalidate = options.revalidate ?? DEFAULT_REVALIDATE;
    const staleWhileRevalidate = options.staleWhileRevalidate ?? DEFAULT_STALE_WHILE_REVALIDATE;

    return `s-maxage=${revalidate}, stale-while-revalidate=${staleWhileRevalidate}`;
  }

  /**
   * Parse ISR options from pragma comments in source.
   *
   * @param {string[]} comments - Array of comment strings
   * @returns {Object|null} - Parsed options or null if no pragma found
   */
  static parsePragma(comments) {
    for (const comment of comments) {
      const match = comment.match(/Pragma:\s*revalidate\s+(\d+)/i);
      if (match) {
        return { revalidate: parseInt(match[1], 10) };
      }
    }
    return null;
  }

  /**
   * Extract route parameters from request URL.
   *
   * @param {Request} request - The incoming request
   * @returns {Object} - Extracted parameters
   */
  static extractParams(request) {
    const url = new URL(request.url);
    const params = {};

    for (const [key, value] of url.searchParams) {
      params[key] = value;
    }

    return params;
  }
}

// Export constants for external use
export { DEFAULT_REVALIDATE, DEFAULT_STALE_WHILE_REVALIDATE };
